components/raftstore-v2/src/raft/storage.rs

```diff
@@ -240,6 +258,159 @@ impl<ER: RaftEngine> Storage<ER> {
+                "region_id" => self.region_state.get_region().get_id(),
+                "peer_id" => self.peer.get_id(),
+                "snap_epoch" => ?snap_epoch,
+                "latest_epoch" => ?latest_epoch,
+            );
+            STORE_SNAPSHOT_VALIDATION_FAILURE_COUNTER.epoch.inc();
+            return false;
+        }
+
+        true
+    }
+
+    /// Gets a snapshot. Returns `SnapshotTemporarilyUnavailable` if there is no
+    /// unavailable snapshot.
+    pub fn snapshot(&self, request_index: u64, to: u64) -> raft::Result<Snapshot> {
+        let mut snap_state = self.snap_state.borrow_mut();
+
+        let mut tried = false;
+        let mut last_canceled = false;
+        if let SnapState::Generating {
+            ref canceled,
+            ref receiver,
+            ..
+        } = *snap_state
+        {
+            tried = true;
+            last_canceled = canceled.load(Ordering::SeqCst);
+            match receiver.try_recv() {
+                Err(TryRecvError::Empty) => {
+                    return Err(raft::Error::Store(
+                        raft::StorageError::SnapshotTemporarilyUnavailable,
+                    ));
+                }
+                Ok(s) if !last_canceled => {
+                    *snap_state = SnapState::Relax;
+                    if self.validate_snap(&s, request_index) {
+                        return Ok(s);
+                    }
+                }
+                Err(TryRecvError::Disconnected) | Ok(_) => {
+                    *snap_state = SnapState::Relax;
+                    warn!(
+                        self.logger(),
+                        "failed to try generating snapshot";
+                        "region_id" => self.region().get_id(),
+                        "peer_id" => self.peer().get_id(),
+                        "request_peer" => to,
+                    );
+                }
+            }
+        }
+
+        if SnapState::Relax != *snap_state {
+            panic!(
+                "[region {}] unexpected state: {:?}",
+                self.region().get_id(),
+                *snap_state
+            );
+        }
+
+        if !tried || !last_canceled {
+            STORE_SNAPSHOT_RETIRES_COUNTER.inc();
+        }
+
+        info!(
+            self.logger(),
+            "requesting snapshot";
+            "region_id" => self.region().get_id(),
+            "peer_id" => self.peer().get_id(),
+            "tablet_index" => self.get_tablet_index(),
+            "request_index" => request_index,
+            "request_peer" => to,
+        );
+
+        let (sender, receiver) = mpsc::sync_channel(1);
+        // TODO(nolouch): remove cancel?
+        let canceled = Arc::new(AtomicBool::new(false));
+        let index = Arc::new(AtomicU64::new(0));
+        *snap_state = SnapState::Generating {
+            canceled: canceled.clone(),
+            index,
+            receiver,
+        };
+
+        let store_id = self
```

Comment:
What's the purpose?

Reply:
I mistake here, that was used for tiflash  with `multi_sst_sanpshot` in v1.